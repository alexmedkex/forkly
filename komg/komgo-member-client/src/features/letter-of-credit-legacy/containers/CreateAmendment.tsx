import { Divider, Header, Segment } from 'semantic-ui-react'
import Ajv from 'ajv'
import * as React from 'react'
import { compose } from 'redux'
import { AMENDMENT_BASE_SCHEMA, ILCAmendmentBase, IDiff, ITrade, ICargo } from '@komgo/types'
import { withPermissions, WithPermissionsProps } from '../../../components/with-permissions'
import { withLoaders, WithLoaderProps } from '../../../components/with-loaders'
import { RouteComponentProps, withRouter } from 'react-router'
import { connect } from 'react-redux'
import { ApplicationState } from '../../../store/reducers'
import { TradeActionType } from '../../trades/store/types'
import { ILetterOfCredit } from '../types/ILetterOfCredit'
import { LetterOfCreditActionType } from '../store/types'
import { ErrorMessage } from '../../../components/error-message'
import { getLetterOfCreditWithTradeAndMovements } from '../store/actions'
import { submitLetterOfCreditAmendment } from '../store/amendments/actions'
import { LoadingTransition } from '../../../components/loading-transition'
import { cargoDiff, tradeDiff } from '../utils/DiffUtils'
import { Unauthorized, Wizard } from '../../../components'
import { CreateAmendmentStateMachine, Step } from '../components/amendment/CreateAmendmentFlow'
import TradeStep from '../components/amendment/TradeStep'
import SummaryStep from '../components/amendment/SummaryStep'
import LetterOfCreditAmendmentStep, { amendableFields } from '../components/amendment/LetterOfCreditAmendmentStep'
import { TRANSITION } from '../../../components/wizard'
import { tradeFinanceManager } from '@komgo/permissions'
import { applyPatch } from 'fast-json-patch'
// TODO change to lc schema in komgo-types
import LETTER_OF_CREDIT_SCHEMA from '../schemas/letter-of-credit.schema.json'
import { toFormikErrors } from '../../../utils/validator'
import { FormikErrors, FormikValues } from 'formik'
import { fieldToLabel } from '../constants/fieldsByStep'
import { pathToKey, keyToPath } from '../components/amendment/PropertyEditor'
import { loadingSelector } from '../../../store/common/selectors'
import { findErrors } from '../../../store/common/selectors/errorSelector'
import { ServerError } from '../../../store/common/types'
import { LetterOfCreditAmendmentActionType } from '../store/amendments/types'

const validator = new Ajv({ allErrors: true }).addSchema(LETTER_OF_CREDIT_SCHEMA).addSchema(AMENDMENT_BASE_SCHEMA)
const lcValidationSchemaKeyRef = 'http://komgo.io/letter-of-credit'

export const findDiffsByTypes = (diffs: IDiff[], types: string[]) => diffs.filter(d => types.includes(d.type))

export interface CreateAmendmentProps extends WithLoaderProps, WithPermissionsProps, RouteComponentProps<any> {
  tradeId: string
  letterOfCreditId: string
  initialValues: ILCAmendmentBase
  letterOfCredit: ILetterOfCredit
  getLetterOfCreditWithTradeAndMovements: (params: { id: string }) => null
  getTrade: (params: { id: string }) => null
  submitLetterOfCreditAmendment: (values: ILCAmendmentBase, lcStaticId: string) => any
  submitAmendmentLoading: boolean
  submitAmendmentError: ServerError
}

export interface CreateAmendmentState {
  values: ILCAmendmentBase
}

export const LetterOfCreditAmendmentContext = React.createContext({})

export class CreateAmendment extends React.Component<CreateAmendmentProps, CreateAmendmentState> {
  constructor(props) {
    super(props)
    this.onSubmit = this.onSubmit.bind(this)
    this.onNext = this.onNext.bind(this)
  }

  componentDidMount(): void {
    this.props.getLetterOfCreditWithTradeAndMovements({ id: this.props.letterOfCreditId })
  }

  onSubmit(values: ILCAmendmentBase) {
    this.props.submitLetterOfCreditAmendment(values, this.props.letterOfCreditId)
  }

  onNext(
    currentValues: ILCAmendmentBase,
    step: string,
    setFieldValue: (field: keyof ILCAmendmentBase, value: any) => void
  ): FormikErrors<FormikValues> {
    const { letterOfCredit } = this.props

    let errors = {}
    if (step === Step.LetterOfCredit) {
      const lcAmendments = findDiffsByTypes(currentValues.diffs, ['ILC'])
      const otherAmendments = findDiffsByTypes(currentValues.diffs, ['ITrade', 'ICargo'])

      // do this first so we don't sanitize any 'magically added' fields which cleanupLetterOfCreditAmendments does...
      const sanitizationErrors = amendmentSanitizationErrors(lcAmendments)
      const actualLcUpdates = buildLetterOfCreditDiffs(lcAmendments)
      const lcSchemaErrors = lcSchemaValidationErrors(letterOfCredit, actualLcUpdates)

      errors = {
        ...sanitizationErrors,
        ...lcSchemaErrors
      }

      if (Object.keys(errors).length === 0) {
        setFieldValue('diffs', [...otherAmendments, ...actualLcUpdates])
      }
    }
    return errors
  }

  render() {
    const { isAuthorized, letterOfCredit, submitAmendmentLoading, errors, submitAmendmentError } = this.props
    if (!isAuthorized(tradeFinanceManager.canManageLCRequests)) {
      return <Unauthorized />
    }

    const [error] = errors
    if (error) {
      return <ErrorMessage title="Amendment Request" error={error} />
    }

    const initialStep = this.props.initialValues.diffs.length > 0 ? Step.Trade : Step.LetterOfCredit
    const stateMachine = CreateAmendmentStateMachine().transitionTo(TRANSITION.GOTO, { step: initialStep })
    return (
      <React.Fragment>
        <Header>Amendment Request</Header>
        <p>Letter Of Credit / {<em>{this.props.letterOfCredit.reference}</em>}</p>
        <Divider />

        {this.props.isFetching ? (
          <Segment basic={true} padded={true}>
            <LoadingTransition title="Loading Amendment data" />
          </Segment>
        ) : (
          <LetterOfCreditAmendmentContext.Provider value={letterOfCredit}>
            {submitAmendmentError && <ErrorMessage title={'Amendment submission error'} error={submitAmendmentError} />}
            <Wizard
              initialValues={this.props.initialValues}
              onSubmit={this.onSubmit}
              validator={validator}
              validationSchemaKeyRef={'http://komgo.io/schema/lc/amendment/1/base'}
              fieldToLabel={() => ''}
              initialStateMachine={stateMachine}
              onNext={this.onNext}
              submitting={submitAmendmentLoading}
            >
              <Wizard.Page step={Step.Trade}>
                <TradeStep />
              </Wizard.Page>
              <Wizard.Page step={Step.LetterOfCredit}>
                <LetterOfCreditAmendmentStep />
              </Wizard.Page>
              <Wizard.Page step={Step.Summary}>
                <SummaryStep />
              </Wizard.Page>
            </Wizard>
          </LetterOfCreditAmendmentContext.Provider>
        )}
      </React.Fragment>
    )
  }
}

const amendmentSanitizationErrors = (input: IDiff[]) => {
  const unchangedFields = input.filter(i => i.oldValue === i.value)

  if (unchangedFields.length !== 0) {
    return unchangedFields.map(unchangedFieldsToErrors).reduce(toObject, {})
  }
  return {}
}

const lcSchemaValidationErrors = (lc: ILetterOfCredit, input: IDiff[]) => {
  const newLetterOfCredit = applyPatch(lc, input as any).newDocument

  if (!validator.validate(lcValidationSchemaKeyRef, newLetterOfCredit)) {
    const currentKeys = input.map(a => pathToKey(a.path))

    const errorsOnAmendableFields = Object.entries(toFormikErrors(validator.errors)).filter(
      ([fieldName]) => currentKeys.includes(fieldName) && amendableFields.includes(fieldName as keyof ILetterOfCredit)
    )

    if (errorsOnAmendableFields.length > 0) {
      return errorsOnAmendableFields.map(replaceFieldNameWithLabel).reduce(toObject, {})
    }
  }
  return {}
}

const replaceFieldNameWithLabel = ([fieldName, message]: [string, string]) => [
  fieldName,
  message.replace(fieldName, fieldToLabel(fieldName))
]
const unchangedFieldsToErrors = field =>
  field.path === ''
    ? ['', 'Empty update field(s) are present.']
    : [
        pathToKey(field.path),
        `'${fieldToLabel(pathToKey(field.path))}' field was left unchanged. Please remove the update or make a change.`
      ]
const toObject = (memo, [fieldName, message]) => ({
  ...memo,
  [fieldName]: message
})

const buildLetterOfCreditDiffs = (input: IDiff[]) => {
  const expiryPlacePath = keyToPath('expiryPlace')

  const availableWithAmendment = input.find(u => u.path === '/availableWith')
  const expiryPlaceAmendment = input.find(u => u.path === expiryPlacePath)

  if (availableWithAmendment) {
    if (expiryPlaceAmendment) {
      return input.map(u => (u.path === expiryPlacePath ? { ...u, value: availableWithAmendment.value } : u))
    } else {
      return [...input, { ...availableWithAmendment, path: expiryPlacePath }]
    }
  } else if (expiryPlaceAmendment) {
    return input.filter(u => u.path !== expiryPlacePath)
  }
  return input.filter(i => i.path !== '')
}

const mapStateToProps = (state: ApplicationState, ownProps: CreateAmendmentProps) => {
  const letterOfCreditId = ownProps.match.params.id
  const letterOfCredit: ILetterOfCredit =
    state
      .get('lettersOfCredit')
      .get('byId')
      .toJS()[letterOfCreditId] || {}

  if (ownProps.isFetching) {
    return {
      letterOfCreditId,
      letterOfCredit,
      initialValues: {
        diffs: []
      }
    }
  }

  const submitAmendmentLoading = loadingSelector(
    state.get('loader').get('requests'),
    [LetterOfCreditAmendmentActionType.SUBMIT_AMENDMENT_REQUEST],
    false
  )

  const [submitAmendmentError] = findErrors(state.get('errors').get('byAction'), [
    LetterOfCreditAmendmentActionType.SUBMIT_AMENDMENT_REQUEST
  ])

  const tradeAndCargoSnapshot: any = letterOfCredit.tradeAndCargoSnapshot || { trade: {}, cargo: {} }

  const newTrade: ITrade = state
    .get('trades')
    .get('trades')
    .toJS()[tradeAndCargoSnapshot.trade._id]

  const [newCargo]: ICargo[] = state
    .get('trades')
    .get('tradeMovements')
    .toJS()
    .filter(cargo => cargo._id === tradeAndCargoSnapshot.cargo._id)

  const oldTrade = tradeAndCargoSnapshot.trade
  const oldCargo = tradeAndCargoSnapshot.cargo

  const tradeDiffs = tradeDiff(oldTrade, newTrade)
  const cargoDiffs = cargoDiff(oldCargo, newCargo)

  const initialValues: ILCAmendmentBase = {
    diffs: [...tradeDiffs, ...cargoDiffs],
    version: 1,
    lcStaticId: letterOfCreditId,
    lcReference: letterOfCredit.reference
  }

  return {
    letterOfCreditId,
    letterOfCredit,
    initialValues,
    submitAmendmentLoading,
    submitAmendmentError
  }
}

export default compose<any>(
  withLoaders({
    actions: [
      LetterOfCreditActionType.LETTER_OF_CREDIT_REQUEST,
      TradeActionType.TRADE_REQUEST,
      TradeActionType.TRADE_MOVEMENTS_REQUEST
    ]
  }),
  withPermissions,
  withRouter,
  connect(mapStateToProps, {
    getLetterOfCreditWithTradeAndMovements,
    submitLetterOfCreditAmendment
  })
)(CreateAmendment)
