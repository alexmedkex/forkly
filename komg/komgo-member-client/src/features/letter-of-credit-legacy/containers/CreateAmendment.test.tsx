import * as React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import { CreateAmendment, CreateAmendmentProps } from './CreateAmendment'
import * as renderer from 'react-test-renderer'
import { fakeCargo, fakeLetterOfCredit, fakeParcel, fakeTrade, fakeLetterOfCreditDiff } from '../utils/faker'
import { createMemoryHistory } from 'history'
import { cargoDiff, tradeDiff } from '../utils/DiffUtils'
import { Step } from '../components/amendment/CreateAmendmentFlow'
import { mount, ReactWrapper } from 'enzyme'
import { AVAILABLE_WITH_OPTIONS } from '../constants'
import { buildFakeAmendmentBase } from '@komgo/types'

const setFieldValue = jest.fn()
const submitLetterOfCreditAmendment = jest.fn()
const letterOfCredit = fakeLetterOfCredit()

describe('CreateAmendment', () => {
  let props: CreateAmendmentProps
  beforeEach(() => {
    props = {
      submitLetterOfCreditAmendment,
      submitAmendmentError: undefined,
      submitAmendmentLoading: false,
      isAuthorized: () => true,
      tradeId: '123',
      letterOfCreditId: letterOfCredit._id,
      initialValues: {
        diffs: [],
        lcReference: 'ref',
        lcStaticId: letterOfCredit._id,
        version: 1
      },
      letterOfCredit,
      getLetterOfCreditWithTradeAndMovements: jest.fn(),
      getTrade: jest.fn(),
      errors: [],
      isFetching: false,
      history: createMemoryHistory(),
      location: {
        pathname: '',
        search: '',
        state: '',
        hash: ''
      },
      match: {
        isExact: false,
        path: '',
        url: '',
        params: { id: 'E1243' }
      },
      staticContext: undefined
    }
  })

  describe('renders', () => {
    it('without tradeFinanceManager.canManageLCRequests permission', () => {
      expect(
        renderer
          .create(
            <Router>
              <CreateAmendment {...props} isAuthorized={() => false} />
            </Router>
          )
          .toJSON()
      ).toMatchSnapshot()
    })
    it('without data', () => {
      expect(
        renderer
          .create(
            <Router>
              <CreateAmendment {...props} />
            </Router>
          )
          .toJSON()
      ).toMatchSnapshot()
    })

    it('with trade diffs', () => {
      const tradeOld = fakeTrade()
      const tradeNew = fakeTrade({ price: 100 })
      const tradeAmendments = tradeDiff(tradeOld, tradeNew)
      expect(
        renderer
          .create(
            <Router>
              <CreateAmendment
                {...props}
                initialValues={{
                  ...props.initialValues,
                  diffs: tradeAmendments
                }}
              />
            </Router>
          )
          .toJSON()
      ).toMatchSnapshot()
    })

    it('with nested cargo diffs', () => {
      const parcel = fakeParcel()
      const cargoOld = fakeCargo({ parcels: [parcel] })
      const cargoNew = fakeCargo({ parcels: [{ ...parcel, vesselName: 'boom!' }] })
      const tradeAmendments = cargoDiff(cargoOld, cargoNew)
      expect(
        renderer
          .create(
            <Router>
              <CreateAmendment
                {...props}
                initialValues={{
                  ...props.initialValues,
                  diffs: tradeAmendments
                }}
              />
            </Router>
          )
          .toJSON()
      ).toMatchSnapshot()
    })

    it('adding a parcel', () => {
      const cargoOld = fakeCargo()
      const cargoNew = fakeCargo({ parcels: [fakeParcel()] })
      const tradeAmendments = cargoDiff(cargoOld, cargoNew)
      expect(
        renderer
          .create(
            <Router>
              <CreateAmendment
                {...props}
                initialValues={{
                  ...props.initialValues,
                  diffs: tradeAmendments
                }}
              />
            </Router>
          )
          .toJSON()
      ).toMatchSnapshot()
    })
  })
  describe('onNext', () => {
    let wrapper: ReactWrapper
    beforeEach(() => {
      jest.resetAllMocks()
      wrapper = mount(
        <Router>
          <CreateAmendment {...props} />
        </Router>
      )
    })

    it('sets expiryPlace if availableWith is changed', () => {
      const amendments = [
        fakeLetterOfCreditDiff({
          path: '/availableWith',
          oldValue: props.letterOfCredit.availableWith,
          value: AVAILABLE_WITH_OPTIONS.ADVISING_BANK
        })
      ]

      const createAmendmentInstance = wrapper.find(CreateAmendment).instance() as CreateAmendment
      const errs = createAmendmentInstance.onNext(
        { ...props.initialValues, diffs: amendments },
        Step.LetterOfCredit,
        setFieldValue
      )

      expect(errs).toEqual({})

      expect(setFieldValue).toHaveBeenCalledWith('diffs', [
        amendments[0],
        {
          op: 'replace',
          path: '/expiryPlace',
          value: amendments[0].value,
          oldValue: amendments[0].oldValue,
          type: 'ILC'
        }
      ])
    })
    it('removes expiryPlace if availableWith is not there', () => {
      const amendments = [fakeLetterOfCreditDiff({ path: '/expiryPlace' })]

      const createAmendmentInstance = wrapper.find(CreateAmendment).instance() as CreateAmendment

      createAmendmentInstance.onNext({ ...props.initialValues, diffs: amendments }, Step.LetterOfCredit, setFieldValue)

      expect(setFieldValue).toHaveBeenCalledWith('diffs', [])
    })
    it('edits expiryPlace if it is already there', () => {
      const amendments = [
        fakeLetterOfCreditDiff({
          path: '/expiryPlace',
          value: AVAILABLE_WITH_OPTIONS.ISSUING_BANK,
          oldValue: AVAILABLE_WITH_OPTIONS.ADVISING_BANK
        }),
        fakeLetterOfCreditDiff({
          path: '/availableWith',
          value: AVAILABLE_WITH_OPTIONS.ADVISING_BANK,
          oldValue: AVAILABLE_WITH_OPTIONS.ISSUING_BANK
        })
      ]

      const createAmendmentInstance = wrapper.find(CreateAmendment).instance() as CreateAmendment

      const errors = createAmendmentInstance.onNext(
        { ...props.initialValues, diffs: amendments },
        Step.LetterOfCredit,
        setFieldValue
      )

      expect(errors).toEqual({})

      expect(setFieldValue).toHaveBeenCalledWith('diffs', [
        {
          op: 'replace',
          path: '/expiryPlace',
          oldValue: AVAILABLE_WITH_OPTIONS.ADVISING_BANK,
          value: AVAILABLE_WITH_OPTIONS.ADVISING_BANK,
          type: 'ILC'
        },
        {
          op: 'replace',
          path: '/availableWith',
          oldValue: AVAILABLE_WITH_OPTIONS.ISSUING_BANK,
          value: AVAILABLE_WITH_OPTIONS.ADVISING_BANK,
          type: 'ILC'
        }
      ])
    })
    it('returns an error if an empty field is present', () => {
      const amendments = [fakeLetterOfCreditDiff({ path: '', value: '', oldValue: '' })]

      const createAmendmentInstance = wrapper.find(CreateAmendment).instance() as CreateAmendment

      const errors = createAmendmentInstance.onNext(
        { ...props.initialValues, diffs: amendments },
        Step.LetterOfCredit,
        setFieldValue
      )

      expect(errors).toEqual({ '': 'Empty update field(s) are present.' })
    })
    it('returns an error if a field oldValue is equal to new value', () => {
      const amendments = [
        fakeLetterOfCreditDiff({
          path: '/availableWith',
          value: AVAILABLE_WITH_OPTIONS.ISSUING_BANK,
          oldValue: AVAILABLE_WITH_OPTIONS.ISSUING_BANK
        })
      ]

      const createAmendmentInstance = wrapper.find(CreateAmendment).instance() as CreateAmendment

      const errors = createAmendmentInstance.onNext(
        { ...props.initialValues, diffs: amendments },
        Step.LetterOfCredit,
        setFieldValue
      )

      expect(errors).toEqual({
        availableWith: "'Available with' field was left unchanged. Please remove the update or make a change."
      })
    })
    it('returns an error if an amended LC field change is invalid', () => {
      const amendments = [
        fakeLetterOfCreditDiff({
          path: '/amount',
          value: -100,
          oldValue: props.letterOfCredit.amount
        })
      ]
      const createAmendmentInstance = wrapper.find(CreateAmendment).instance() as CreateAmendment

      const errors = createAmendmentInstance.onNext(
        { ...props.initialValues, diffs: amendments },
        Step.LetterOfCredit,
        setFieldValue
      )

      expect(errors).toEqual({
        amount: "'Opening amount' should be greater than or equal to 0.01"
      })
    })
  })
  describe('submission', () => {
    let wrapper: ReactWrapper
    beforeEach(() => {
      jest.resetAllMocks()
      wrapper = mount(
        <Router>
          <CreateAmendment {...props} />
        </Router>
      )
    })
    it('calls submitLetterOfCreditAmendment with right args', () => {
      const createAmendmentInstance = wrapper.find(CreateAmendment).instance() as CreateAmendment

      const amendment = buildFakeAmendmentBase()
      createAmendmentInstance.onSubmit(amendment)

      expect(submitLetterOfCreditAmendment).toHaveBeenCalledWith(amendment, letterOfCredit._id)
    })
  })
})
