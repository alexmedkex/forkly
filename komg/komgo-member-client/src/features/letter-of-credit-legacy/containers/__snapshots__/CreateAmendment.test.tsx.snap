// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAmendment renders adding a parcel 1`] = `
Array [
  <div
    className="ui header"
  >
    Amendment Request
  </div>,
  <p>
    Letter Of Credit / 
    <em>
      LC18-MER-1
    </em>
  </p>,
  <div
    className="ui divider"
  />,
  <div
    className="ui horizontal list"
    role="list"
  >
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#5700b5",
          "fontWeight": "bold",
        }
      }
    >
      Trade
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#1c2936",
          "cursor": "undefined",
          "fontWeight": "bold",
        }
      }
    >
      Letter of credit
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#1c2936",
          "cursor": "undefined",
          "fontWeight": "bold",
        }
      }
    >
      Summary
    </div>
  </div>,
  <div
    className="ui hidden divider"
  />,
  .c0.c0.c0 {
  text-transform: uppercase;
  border: none;
  background-color: #f2f5f8;
}

<form
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="ui block header c0"
    >
      Updates
    </div>
    <div
      className="ui basic segment"
    >
      <div>
        <strong
          className="ui checked checkbox"
          onChange={[Function]}
          onClick={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "margin": 0,
            }
          }
        >
          <input
            checked={true}
            className="hidden"
            name="parcels.0"
            onClick={[Function]}
            readOnly={true}
            tabIndex={0}
            type="checkbox"
          />
          <label>
            Parcels 0
          </label>
        </strong>
        <div
          className="ui list"
          role="list"
          style={
            Object {
              "margin": "0px 0px 10px 30px",
              "padding": 0,
            }
          }
        >
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
            style={
              Object {
                "padding": "5px 0 0 0",
              }
            }
          >
            <div
              className="content"
            >
              <div
                className="description"
                data-test-id="ICargo/parcels/0-current"
                style={
                  Object {
                    "fontSize": "0.9em",
                    "padding": 0,
                  }
                }
              >
                Current Value:
                <span
                  style={
                    Object {
                      "color": "#46484a",
                      "whiteSpace": "pre",
                    }
                  }
                >
                   
                  -
                </span>
              </div>
            </div>
          </div>
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
            style={
              Object {
                "padding": "5px 0 0 0",
              }
            }
          >
            <div
              className="content"
            >
              <div
                className="description"
                data-test-id="ICargo/parcels/0-new"
                style={
                  Object {
                    "fontSize": "0.9em",
                    "padding": 0,
                  }
                }
              >
                New Value:
                <span
                  style={
                    Object {
                      "color": "#009ea8",
                      "whiteSpace": "pre",
                    }
                  }
                >
                   
                  
  id: 1234
  laycanPeriod: 
    startDate: 2019/01/03
    endDate: 2019/01/03
  modeOfTransport: VESSEL
  vesselIMO: 1
  vesselName: Test
  loadingPort: Test
  dischargeArea: Test
  inspector: Test
  deemedBLDate: 2019/01/03
  quantity: 1

                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ui divider"
    />
    <div
      className="ui basic right aligned segment"
    >
      <button
        className="ui button"
        name="next"
        onClick={[Function]}
        role="button"
        type="button"
      >
        Next
      </button>
    </div>
  </form>,
]
`;

exports[`CreateAmendment renders with nested cargo diffs 1`] = `
Array [
  <div
    className="ui header"
  >
    Amendment Request
  </div>,
  <p>
    Letter Of Credit / 
    <em>
      LC18-MER-1
    </em>
  </p>,
  <div
    className="ui divider"
  />,
  <div
    className="ui horizontal list"
    role="list"
  >
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#5700b5",
          "fontWeight": "bold",
        }
      }
    >
      Trade
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#1c2936",
          "cursor": "undefined",
          "fontWeight": "bold",
        }
      }
    >
      Letter of credit
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#1c2936",
          "cursor": "undefined",
          "fontWeight": "bold",
        }
      }
    >
      Summary
    </div>
  </div>,
  <div
    className="ui hidden divider"
  />,
  .c0.c0.c0 {
  text-transform: uppercase;
  border: none;
  background-color: #f2f5f8;
}

<form
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="ui block header c0"
    >
      Updates
    </div>
    <div
      className="ui basic segment"
    >
      <div>
        <strong
          className="ui checked checkbox"
          onChange={[Function]}
          onClick={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "margin": 0,
            }
          }
        >
          <input
            checked={true}
            className="hidden"
            name="parcels.0.vesselName"
            onClick={[Function]}
            readOnly={true}
            tabIndex={0}
            type="checkbox"
          />
          <label>
            Parcels 0 Vessel Name
          </label>
        </strong>
        <div
          className="ui list"
          role="list"
          style={
            Object {
              "margin": "0px 0px 10px 30px",
              "padding": 0,
            }
          }
        >
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
            style={
              Object {
                "padding": "5px 0 0 0",
              }
            }
          >
            <div
              className="content"
            >
              <div
                className="description"
                data-test-id="ICargo/parcels/0/vesselName-current"
                style={
                  Object {
                    "fontSize": "0.9em",
                    "padding": 0,
                  }
                }
              >
                Current Value:
                <span
                  style={
                    Object {
                      "color": "#46484a",
                      "whiteSpace": "pre",
                    }
                  }
                >
                   
                  Test
                </span>
              </div>
            </div>
          </div>
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
            style={
              Object {
                "padding": "5px 0 0 0",
              }
            }
          >
            <div
              className="content"
            >
              <div
                className="description"
                data-test-id="ICargo/parcels/0/vesselName-new"
                style={
                  Object {
                    "fontSize": "0.9em",
                    "padding": 0,
                  }
                }
              >
                New Value:
                <span
                  style={
                    Object {
                      "color": "#009ea8",
                      "whiteSpace": "pre",
                    }
                  }
                >
                   
                  boom!
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ui divider"
    />
    <div
      className="ui basic right aligned segment"
    >
      <button
        className="ui button"
        name="next"
        onClick={[Function]}
        role="button"
        type="button"
      >
        Next
      </button>
    </div>
  </form>,
]
`;

exports[`CreateAmendment renders with trade diffs 1`] = `
Array [
  <div
    className="ui header"
  >
    Amendment Request
  </div>,
  <p>
    Letter Of Credit / 
    <em>
      LC18-MER-1
    </em>
  </p>,
  <div
    className="ui divider"
  />,
  <div
    className="ui horizontal list"
    role="list"
  >
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#5700b5",
          "fontWeight": "bold",
        }
      }
    >
      Trade
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#1c2936",
          "cursor": "undefined",
          "fontWeight": "bold",
        }
      }
    >
      Letter of credit
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#1c2936",
          "cursor": "undefined",
          "fontWeight": "bold",
        }
      }
    >
      Summary
    </div>
  </div>,
  <div
    className="ui hidden divider"
  />,
  .c0.c0.c0 {
  text-transform: uppercase;
  border: none;
  background-color: #f2f5f8;
}

<form
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="ui block header c0"
    >
      Updates
    </div>
    <div
      className="ui basic segment"
    >
      <div>
        <strong
          className="ui checked checkbox"
          onChange={[Function]}
          onClick={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "margin": 0,
            }
          }
        >
          <input
            checked={true}
            className="hidden"
            name="price"
            onClick={[Function]}
            readOnly={true}
            tabIndex={0}
            type="checkbox"
          />
          <label>
            Price
          </label>
        </strong>
        <div
          className="ui list"
          role="list"
          style={
            Object {
              "margin": "0px 0px 10px 30px",
              "padding": 0,
            }
          }
        >
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
            style={
              Object {
                "padding": "5px 0 0 0",
              }
            }
          >
            <div
              className="content"
            >
              <div
                className="description"
                data-test-id="ITrade/price-current"
                style={
                  Object {
                    "fontSize": "0.9em",
                    "padding": 0,
                  }
                }
              >
                Current Value:
                <span
                  style={
                    Object {
                      "color": "#46484a",
                      "whiteSpace": "pre",
                    }
                  }
                >
                   
                  73.14
                </span>
              </div>
            </div>
          </div>
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
            style={
              Object {
                "padding": "5px 0 0 0",
              }
            }
          >
            <div
              className="content"
            >
              <div
                className="description"
                data-test-id="ITrade/price-new"
                style={
                  Object {
                    "fontSize": "0.9em",
                    "padding": 0,
                  }
                }
              >
                New Value:
                <span
                  style={
                    Object {
                      "color": "#009ea8",
                      "whiteSpace": "pre",
                    }
                  }
                >
                   
                  100.00
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ui divider"
    />
    <div
      className="ui basic right aligned segment"
    >
      <button
        className="ui button"
        name="next"
        onClick={[Function]}
        role="button"
        type="button"
      >
        Next
      </button>
    </div>
  </form>,
]
`;

exports[`CreateAmendment renders without data 1`] = `
Array [
  <div
    className="ui header"
  >
    Amendment Request
  </div>,
  <p>
    Letter Of Credit / 
    <em>
      LC18-MER-1
    </em>
  </p>,
  <div
    className="ui divider"
  />,
  <div
    className="ui horizontal list"
    role="list"
  >
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#1c2936",
          "cursor": "undefined",
          "fontWeight": "bold",
        }
      }
    >
      Trade
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#5700b5",
          "fontWeight": "bold",
        }
      }
    >
      Letter of credit
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "color": "#1c2936",
          "cursor": "undefined",
          "fontWeight": "bold",
        }
      }
    >
      Summary
    </div>
  </div>,
  <div
    className="ui hidden divider"
  />,
  .c0.c0.c0 {
  text-transform: uppercase;
  border: none;
  background-color: #f2f5f8;
}

.c1 {
  color: #5700b5;
  border: none;
  cursor: pointer;
  background-color: #ffffff;
}

.c1:focus {
  outline-color: #ffffff;
}

<form
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="ui block header c0"
    >
      Updates
    </div>
    <div
      className="ui basic segment"
    >
      <button
        className="c1"
        color="#5700b5"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        + Add field
      </button>
    </div>
    <div
      className="ui divider"
    />
    <div
      className="ui basic right aligned segment"
    >
      <button
        className="ui button"
        name="previous"
        onClick={[Function]}
        role="button"
        type="button"
      >
        Previous
      </button>
      <button
        className="ui button"
        name="next"
        onClick={[Function]}
        role="button"
        type="button"
      >
        Next
      </button>
    </div>
  </form>,
]
`;

exports[`CreateAmendment renders without tradeFinanceManager.canManageLCRequests permission 1`] = `
<h1>
  You don't have enough permissions to view this page
</h1>
`;
